File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        ### Intuitive idea is to use mergeSort algo in
        ### this list of link_Lists.

        # 1- merge two linkedList, from Lists, at the time.
        # and the result of this two linkedlist merge to be
        # stored in one of the two list.
        amountOfList = len(lists)
        numOfList = 1

        # In each while itiration we mergeSort each 
        # two linkedlists... until we merge them 
        # all in list[0]
        while numOfList < amountOfList:
            for i in range(0, amountOfList - numOfList, numOfList * 2):
                lists[i] = self.mergeSortTwoList(lists[i], lists[i + numOfList])
            numOfList *= 2

        return lists[0] if amountOfList > 0 else None

    ### MergeSortTwoList sort for two sorted link list.
    def mergeSortTwoList(self, l1, l2):
        head = node = ListNode(0)

        while l1 and l2:
            if l1.val <= l2.val:
                node.next = l1
                l1 = l1.next
            else:
                node.next = l2
                ## This two steps below are hard steps to 
                ## think about them during interview.
                # So, since we are merging to two list 
                # into list[i], then we need to point l2 
                # to the remaining of l1 linked list. And then
                # point l1 to the node.next.next where 
                # we stind now  
                l2 = l1
                l1 = node.next.next
            node = node.next

        if not l1:
            node.next = l2
        else:
            node.next = l1
        
        return head.next
            

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.go
"""
[
  1->4->5,
  1->3->4,
  2->6
]
- B.F: 
  Merge linked list by linked list. The Time complexity O(k*n)

To avoid to repete the merging each linked list and improve the 
time complexity we can merge two linkedList by two linkedList: 
   example:
    - merge 1->4->5 and 1->3->4 ===> 1->1->3->4->4->5
    - merge 1->1->3->4->4->5 and 2->6

Approach:
- Merge each two linked list i and i+1 and 
  append them in new mergerList (list of merged linked lists).
- return the head of final merged list
- build the merge function of two linked list
- Time comlexity is n*log(k), since we keep dividing the lists by 2 logk times
  Space complexity is: O(k)

"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if lists is None or len(lists) == 0:
            return None

        while len(lists) > 1:
            mergeList = []
            for i in range(0, len(lists), 2):
                l1 = lists[i]
                l2 = lists[i+1] if i+1 < len(lists) else None
                mergeList.append(self.mergeTwoList(l1, l2))
            lists = mergeList

        return lists[0]

    def mergeTwoList(self, l1, l2):
        dummy = ListNode()
        tail = dummy
        
        while l1 and l2:
            if l1.val < l2.val:
                tail.next = l1
                l1 = l1.next
            else:
                tail.next = l2
                l2 = l2.next
            # tail.next already filled, move to the tail.next
            tail = tail.next

        if l1:
            tail.next = l1
        if l2:
            tail.next = l2

        return dummy.next